ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,525 @@
+		}
+
+		return NULL;
+	}
+
+	if (uriparser_base_urls == NULL) {
+		return uriparser_create_uris(uriparser_uri, original_uri_str, NULL, NULL);
+	}
+
+	UriUriA *uriparser_base_url = uriparser_copy_uri(uriparser_base_urls->uri);
+
+	UriUriA *absolute_uri = emalloc(sizeof(UriUriA));
+
+	if (uriAddBaseUriExMmA(absolute_uri, uriparser_uri, uriparser_base_url, URI_RESOLVE_STRICTLY, &uriparser_memory_manager) != URI_SUCCESS) {
+		zend_string_release_ex(original_uri_str, false);
+		uriFreeUriMembersMmA(uriparser_uri, &uriparser_memory_manager);
+		efree(uriparser_uri);
+		uriFreeUriMembersMmA(uriparser_base_url, &uriparser_memory_manager);
+		efree(uriparser_base_url);
+		efree(absolute_uri);
+
+		if (!silent) {
+			throw_invalid_uri_exception();
+		}
+
+		return NULL;
+	}
+
+	/* TODO fix freeing: if the following code runs, then we'll have use-after-free-s because uriparser doesn't
+	   copy the input. If we don't run the following code, then we'll have memory leaks...
+	uriFreeUriMembersMmA(uriparser_base_url, &uriparser_memory_manager);
+	efree(uriparser_base_url);
+	uriFreeUriMembersMmA(uriparser_uri, &uriparser_memory_manager);
+	efree(uriparser_uri);
+	 */
+
+	return uriparser_create_uris(absolute_uri, original_uri_str, NULL, NULL);
+}
+
+void *uriparser_parse_uri(const zend_string *uri_str, const void *base_url, zval *errors, bool silent)
+{
+	return uriparser_parse_uri_ex(uri_str, base_url, silent);
+}
+
+static void *uriparser_clone_uri(void *uri)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) uri;
+
+	return uriparser_create_uris(
+		uriparser_copy_uri(uriparser_uris->uri),
+		zend_string_copy(uriparser_uris->uri_str),
+		uriparser_uris->normalized_uri != NULL ? uriparser_copy_uri(uriparser_uris->normalized_uri) : NULL,
+		uriparser_uris->normalized_uri_str != NULL ? zend_string_copy(uriparser_uris->normalized_uri_str) : NULL
+	);
+}
+
+static zend_string *uriparser_uri_to_string(void *uri, uri_recomposition_mode_t recomposition_mode, bool exclude_fragment)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) uri;
+	UriUriA *uriparser_uri = uriparser_uris->uri;
+
+	if ((recomposition_mode == URI_RECOMPOSITION_NORMALIZED_UNICODE || recomposition_mode == URI_RECOMPOSITION_NORMALIZED_ASCII) &&
+		uriparser_uris->normalized_uri == NULL
+	) {
+		uriparser_uris->normalized_uri = uriparser_copy_uri(uriparser_uris->uri);
+		if (uriparser_normalize_uri(uriparser_uris->normalized_uri) == FAILURE) {
+			return NULL;
+		}
+		uriparser_uri = uriparser_uris->normalized_uri;
+	}
+
+	int charsRequired;
+	if (uriToStringCharsRequiredA(uriparser_uri, &charsRequired) != URI_SUCCESS) {
+		return NULL;
+	}
+
+	charsRequired++;
+
+	zend_string *uri_string = zend_string_alloc(charsRequired - 1, false);
+	if (uriToStringA(ZSTR_VAL(uri_string), uriparser_uri, charsRequired, NULL) != URI_SUCCESS) {
+		zend_string_release(uri_string);
+		return NULL;
+	}
+
+	if (exclude_fragment) {
+		char *pos = strrchr(ZSTR_VAL(uri_string), '#');
```

Comment:
This won't work with strings that have NUL bytes; but that's probably impossible here. Is this after encoding?

Reply:
NUL bytes should indeed be impossible, as all the string parameters are `Z_PARAM_PATH_STR`, although it has just come to my mind that probably NUL bytes can be injected during unserialization 🤔  I think I'll have to prevent this case as well.

If the raw representation is used, then there's no encoding, but the # character falls within the "gen-delim" character set, so it can only be part of the URI as a delimiter.